File: packages/kotlin/vidcall-protocol/src/main/kotlin/io/vidcall/protocol/MessageType.kt
package io.vidcall.protocol

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Envelope types, mirrors the `type` enum of `protocol/schema.json`.
 *
 * Wire names are the schema enum values (`screen-share`, `quality-warning`, ...);
 * the Kotlin names are idiomatic upper-camel identifiers.
 */
@Serializable
enum class MessageType {
    @SerialName("join") JOIN,
    @SerialName("leave") LEAVE,
    @SerialName("offer") OFFER,
    @SerialName("answer") ANSWER,
    @SerialName("ice") ICE,
    @SerialName("presence") PRESENCE,
    @SerialName("reaction") REACTION,
    @SerialName("chat") CHAT,
    @SerialName("screen-share") SCREEN_SHARE,
    @SerialName("quality-warning") QUALITY_WARNING,
    @SerialName("sfu") SFU,
    @SerialName("error") ERROR,
    @SerialName("ping") PING,
    @SerialName("pong") PONG,
}

File: packages/dart/lib/src/protocol/message_type.dart
/// Wire `type` values from `protocol/schema.json`.
///
/// The protocol is additive-only: clients must tolerate unknown `type`
/// strings (ignore + log). [`MessageType.tryParse`] returns `null` for those.
enum MessageType {
  join('join'),
  leave('leave'),
  offer('offer'),
  answer('answer'),
  ice('ice'),
  presence('presence'),
  reaction('reaction'),
  chat('chat'),
  screenShare('screen-share'),
  qualityWarning('quality-warning'),
  sfu('sfu'),
  error('error'),
  ping('ping'),
  pong('pong');

  const MessageType(this.wire);

  /// The exact string used on the wire.
  final String wire;

  /// Parses [wire] into a [MessageType], or `null` for unknown
  /// (forward-compatible) types.
  static MessageType? tryParse(String wire) {
    for (final type in MessageType.values) {
      if (type.wire == wire) {
        return type;
      }
    }
    return null;
  }

  @override
  String toString() => wire;
}
